reporting/src/request/index.js

```diff
@@ -212,487 +214,403 @@ export default class RequestMonitor {
-        name: `antitracking.${stage}`,
-        spec: 'blocking',
-        fn: (...args) => this.pipelines[stage].execute(...args),
-      }),
+    chrome.webRequest.onHeadersReceived.addListener(
+      this.onHeadersReceived,
+      { urls },
+      blockingWebRequest
+        ? Object.values(chrome.webRequest.OnHeadersReceivedOptions)
+        : undefined,
     );
-  }
-
-  unloadPipeline() {
-    Object.keys(this.pipelineSteps || {}).forEach((key) => {
-      const step = this.pipelineSteps[key];
-      if (step.unload) {
-        step.unload();
-      }
+    chrome.webRequest.onCompleted.addListener(this.onCompleted, { urls });
+    chrome.webRequest.onErrorOccurred.addListener(this.onErrorOccurred, {
+      urls,
     });
-
-    Object.keys(this.pipelines).forEach((stage) => {
-      this.pipelines[stage].unload();
-    });
-
-    Object.keys(this.pipelines).map((stage) =>
-      this.webRequestPipeline.removePipelineStep(
-        stage,
-        `antitracking.${stage}`,
-      ),
-    );
-    this.pipelines = {};
   }
 
   unload() {
     // Check is active usage, was sent
     this.qs_whitelist.destroy();
-    this.unloadPipeline();
+    this.cookieContext.unload();
+    this.oAuthDetector.unload();
+    this.tokenTelemetry.unload();
+    this.tokenExaminer.unload();
+    this.tokenChecker.unload();
+
+    chrome.webRequest.onBeforeRequest.removeListener(this.onBeforeRequest);
+    chrome.webRequest.onBeforeSendHeaders.removeListener(
+      this.onBeforeSendHeaders,
+    );
+    chrome.webRequest.onHeadersReceived.removeListener(this.onHeadersReceived);
+    chrome.webRequest.onCompleted.removeListener(this.onCompleted);
+    chrome.webRequest.onErrorOccurred.removeListener(this.onErrorOccurred);
+
     this.db.unload();
     this.dayChangedInterval = this.dayChangedInterval.stop();
   }
 
+  onBeforeRequest = (details) => {
+    const state = WebRequestContext.fromDetails(
+      details,
+      this.pageStore,
+      'onBeforeRequest',
+    );
+    const response = new BlockingResponse(details, 'onBeforeRequest');
+    // checkState
+    if (checkValidContext(state) === false) {
+      return response.toWebRequestResponse();
+    }
+    // oAuthDetector.checkMainFrames
+    if (this.oAuthDetector.checkMainFrames(state) === false) {
+      return response.toWebRequestResponse();
+    }
+    // checkIsMainDocument
+    if (!state.isMainFrame === false) {
+      return response.toWebRequestResponse();
+    }
+    // checkSameGeneralDomain
+    if (checkSameGeneralDomain(state) === false) {
+      return response.toWebRequestResponse();
+    }
+    // cancelRecentlyModified
+    if (this.cancelRecentlyModified(state, response) === false) {
+      return response.toWebRequestResponse();
+    }
+    // pageLogger.onBeforeRequest
+    if (this.pageLogger.onBeforeRequest(state) === false) {
+      return response.toWebRequestResponse();
+    }
+    // logIsTracker
+    setTimeout(() => {
```

Comment:
General comment to all uses of setTimeout(..., 1) here: I wonder if it the best way to deal with it, since you mostly lose control afterwards (opening it up to races, for instance). If the intend is only to run it after the processing of the event, setTimeout(..., 0) might also be better.

But I wonder if inviting the potential of interleaving event handling is not introducing avoidable races. For testing, it also makes it more challenging, since you need to bake in low-level assumptions about the clock into the tests. (And some assumption of them might turn out to be unrealistic.)

setTimeout(..., 1) is generally a 4 ms delay on Chrome if I recall correctly (since it rounds up). There can be quite some events during such an interval.

Reply:
This was not changed https://github.com/whotracksme/webextension-packages/blob/6b800a596e6ce9d80bceb59e0f21795aedde75dc/reporting/src/webrequest-pipeline/pipeline.js#L200

Reply:
https://github.com/whotracksme/webextension-packages/pull/117